File: scripts/ingest.ts
import 'dotenv/config';
import fs from 'fs';
import path from 'path';
import { generateEmbedding } from '../src/lib/embeddings';
import { Pool, neonConfig } from '@neondatabase/serverless';
import ws from 'ws';

neonConfig.webSocketConstructor = ws;

async function main() {
  const pool = new Pool({ connectionString: process.env.DATABASE_URL });
  
  console.log('Starting ingestion pipeline...');
  const dataPath = path.join(__dirname, '../data/legalbench.json');
  if (!fs.existsSync(dataPath)) {
    console.error('Dataset not found at', dataPath);
    return;
  }
  
  const qaPairs = JSON.parse(fs.readFileSync(dataPath, 'utf-8'));
  console.log(`Loaded ${qaPairs.length} QA pairs from LegalBench-RAG. Process...`);

  console.log('Altering Chunk table to accept 384-dimensional vectors (Transformers.js)...');
  await pool.query(`ALTER TABLE "Chunk" ALTER COLUMN "embedding" TYPE vector(384);`);

  // Group snippets by document (corpus_file)
  const docsMap = new Map<string, Set<string>>();
  
  for (const pair of qaPairs) {
    const meta = pair.metadata;
    const file = meta.corpus_file;
    if (!docsMap.has(file)) docsMap.set(file, new Set());
    
    // Some QA pairs have multiple snippets
    for (const snippet of meta.snippets) {
       docsMap.get(file)!.add(snippet.answer.trim());
    }
  }
  
  console.log(`Found ${docsMap.size} unique contracts in the dataset sample.`);

  let totalChunks = 0;
  for (const [corpusFile, uniqueSnippets] of docsMap.entries()) {
    console.log(`Processing Document: ${corpusFile}`);
    
    // 1. Create or Find Document
    const documentId = crypto.randomUUID();
    await pool.query(`
      INSERT INTO "Document" ("id", "filename", "title", "source_corpus", "createdAt") 
      VALUES ($1, $2, $3, $4, NOW())
    `, [documentId, corpusFile.split('/').pop() || corpusFile, corpusFile, 'legalbench-rag']);

    // 2. Generate Embeddings & Save Chunks
    for (const text of uniqueSnippets) {
      if (!text) continue;
      
      console.log(` Generating embedding for snippet (${text.length} chars)...`);
      const embeddingArray = await generateEmbedding(text);
      
      await pool.query(`
        INSERT INTO "Chunk" ("id", "document_id", "content", "embedding", "char_start", "char_end") 
        VALUES (gen_random_uuid(), $1, $2, $3::vector, 0, 0)
      `, [documentId, text, JSON.stringify(embeddingArray)]);
      totalChunks++;
    }
  }
  
  console.log(`\n✅ Ingestion Complete! Inserted ${docsMap.size} Documents and ${totalChunks} Chunks.`);
}

main().catch(e => {
  console.error(e);
  process.exit(1);
}).finally(() => {
  process.exit(0);
});

File: python/scripts/ingest.py
"""
Synapse RAG — Ingestion Pipeline (Python/LangChain + Qdrant)

Reads data/legalbench.json, creates LangChain Document objects,
and ingests them into a local Qdrant instance using HuggingFaceEmbeddings.

Usage:
    python scripts/ingest.py
"""

import json
import sys
import uuid
import shutil
from pathlib import Path

# Allow running from project root or from python/
sys.path.insert(0, str(Path(__file__).parent.parent))

from dotenv import load_dotenv

load_dotenv(dotenv_path=Path(__file__).parent.parent.parent / ".env")
load_dotenv()

from langchain_core.documents import Document
from langchain_qdrant import QdrantVectorStore
from app.embeddings import get_embeddings


def main():
    data_path = Path(__file__).parent.parent.parent / "data" / "legalbench.json"
    if not data_path.exists():
        print(f"ERROR: Dataset not found at {data_path}")
        sys.exit(1)

    with open(data_path, "r", encoding="utf-8") as f:
        qa_pairs = json.load(f)

    print(f"Loaded {len(qa_pairs)} QA pairs from LegalBench-RAG.")

    # Group unique snippets by document (corpus_file)
    docs_map: dict[str, set] = {}
    for pair in qa_pairs:
        meta = pair["metadata"]
        file = meta["corpus_file"]
        if file not in docs_map:
            docs_map[file] = set()
        for snippet in meta.get("snippets", []):
            text = snippet.get("answer", "").strip()
            if text:
                docs_map[file].add(text)

    print(f"Found {len(docs_map)} unique contracts in the dataset sample. Preparing documents...")

    langchain_docs = []
    
    for corpus_file, unique_snippets in docs_map.items():
        document_id = str(uuid.uuid4())
        filename = corpus_file.split("/")[-1] or corpus_file

        for text in unique_snippets:
            doc = Document(
                page_content=text,
                metadata={
                    "document_id": document_id,
                    "filename": filename,
                    "source_corpus": "legalbench-rag",
                },
                id=str(uuid.uuid4()),
            )
            langchain_docs.append(doc)

    print(f"Prepared {len(langchain_docs)} chunks for ingestion.")

    qdrant_dir = Path(__file__).parent.parent / "qdrant_db"
    
    # Clear existing DB to prevent duplicates on re-run
    if qdrant_dir.exists():
        print(f"Removing existing Qdrant directory: {qdrant_dir}")
        shutil.rmtree(qdrant_dir)

    print("Initializing Qdrant and computing embeddings (this may take a minute)...")
    embeddings = get_embeddings()
    
    QdrantVectorStore.from_documents(
        documents=langchain_docs,
        embedding=embeddings,
        path=str(qdrant_dir),
        collection_name="synapse_rag",
    )
    
    print(f"\n✅ Ingestion Complete! Inserted {len(langchain_docs)} chunks into Qdrant at {qdrant_dir}.")


if __name__ == "__main__":
    main()
